CenterForOpenScience/osf.io
# File Metadata Backend, API, Tests [PLAT-1128]

## Purpose

This is one big PR to bring together every piece of file metadata that's been merged so far into the feature branch. Truly a group effort!



## Changes

- Add file metadata models
	- FileMetadataRecord
	- FileMetadataSchema
- File Metadata API endpoints
	- List 
	- Create
	- Download
		- downloads the full jsonschema-compatible version of the file metadata, in a specified format
- Serializers for converting file metadata into the format expected by the schema
	- These aren't traditional serializers in the DRF sense
	- They contain the methods to combine the metadata on the file and node with the user entered metadata to create a JSON blob that would validate using the official datacite schema
- New jsonschemas
	- `datacite.json`, taken from a pre-compiled jsonschema [found elsewhere on github](https://github.com/inveniosoftware/datacite/blob/v1.0.1/datacite/schemas/datacite-v4.0.json)
	 - `user_entered_datacite.json` - used for validating what goes in the `metadata` field for `FileMetadataRecord` 
- Process changes in other files:
	- create a new `FileMetadataRecord` for every new file created
	- add a `FILE_METADATA_UPDATED` `NodeLog` every time that happens
	- When a file is copied, its `FileMetadataRecord`s are copied as well, and populated with the old file's `FileMetadataRecord`'s metadata field.
- Housekeeping/organization changes
	- Move files from `api/users/schemas/` to `api/base/schemas/`
- Migration to add a datacite `FileMetadataRecord` for all existing `OsfStorageFile`s

## QA Notes

Surprisingly, not a lot of QA most likely for this gigantic feature as there is nothing yet user facing aside from API endpoints. 

A basic check that the feature was working would be to check that the `metadata_records` relationship appears in file detail endpoints for old files, files can still be created, and that new files have `metadata_records` and detail views on the API.

In particular, here's what I spot checked on staging 3:

A pre-existing node's file list has the metadata_records relationship: https://api.staging3.osf.io/v2/nodes/ma4qn/files/osfstorage/5be5baaca963fe001513516f/
That relationship link can be followed: https://api.staging3.osf.io/v2/files/5be5b7f1a963fe001513515e/metadata_records/
the download link works as expected: https://api.staging3.osf.io/v2/files/5be5b7f1a963fe001513515e/metadata_records/5c54ac6eacbb1a7f958ce75f/download/
A metadata record was also created for a new file https://api.staging3.osf.io/v2/nodes/3jnrd/files/osfstorage/
Tested in Canada and Germany!

pre-existing preprint: https://api.staging3.osf.io/v2/files/5c51a92d2db37d0016cd11af/metadata_records/

Newly created preprint:
https://api.staging3.osf.io/v2/files/5c54b86872a7c20017a66718/metadata_records/

## Documentation

TBD - none exists yet, and we may not want to have user-facing docs for this yet until we are using it ourselves!

## Side Effects

- A new thing is now created every time a file is created.

## Ticket
https://openscience.atlassian.net/browse/PLAT-1128
